### PR TITLE
steam/fhsenv: re-order VK_ICD files

### PR DESCRIPTION
**Pull request by jonringer**
###### Motivation for this change
Order VK_ICD_FILENAMES by preferring more specific drivers
before generic ones.

overlaps with: #126435
may solve:  #126428
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/games/steam/fhsenv.nix [hunk 1]
```diff
@@ -282,7 +282,24 @@ in buildFHSUserEnv rec {
...
+    for prefix in "''${vk_prefixes[@]}"; do
+      for icd_filename in "$prefix"_icd.{x86_64,i686}.json; do
+        icd_filepath=/usr/share/vulkan/icd.d/''${arch_icd}
+        if [ -e "$icd_filepath" ]; then
+          export VK_ICD_FILENAMES=''${VK_ICD_FILENAMES}''${VK_ICD_FILENAMES:+:}"$icd_filepath"
+        fi
+      done
+    done
+
   '' + extraProfile;
```
Comment (SuperSandro2000) [inline pkgs/games/steam/fhsenv.nix:303]:
```suggestion
    done
  '' + extraProfile;
```

File: pkgs/games/steam/fhsenv.nix [hunk 2]
```diff
@@ -282,7 +282,24 @@ in buildFHSUserEnv rec {
...
+      nvidia
+      radeon
+      intel
+      lvp
+    )
+
+    export VK_ICD_FILENAMES=
+    for prefix in "''${vk_prefixes[@]}"; do
+      for icd_filename in "$prefix"_icd.{x86_64,i686}.json; do
+        icd_filepath=/usr/share/vulkan/icd.d/''${arch_icd}
```
Comment (SuperSandro2000) [inline pkgs/games/steam/fhsenv.nix:296]:
```suggestion
        icd_filepath=/usr/share/vulkan/icd.d/$arch_icd
```

File: pkgs/games/steam/fhsenv.nix [hunk 3]
```diff
@@ -282,7 +282,24 @@ in buildFHSUserEnv rec {
...
+        if [ -e "$icd_filepath" ]; then
+          export VK_ICD_FILENAMES=''${VK_ICD_FILENAMES}''${VK_ICD_FILENAMES:+:}"$icd_filepath"
```
Comment (SuperSandro2000) [inline pkgs/games/steam/fhsenv.nix:298]:
```suggestion
          export VK_ICD_FILENAMES=$VK_ICD_FILENAMES''${VK_ICD_FILENAMES:+:}"$icd_filepath"
```

File: pkgs/games/steam/fhsenv.nix (hunk 2 above)
Comment (jonringer) [reply pkgs/games/steam/fhsenv.nix:296]: not sure if I agree, whitespace will cause this to be mis-interpreted

File: pkgs/games/steam/fhsenv.nix (hunk 3 above)
Comment (jonringer) [reply pkgs/games/steam/fhsenv.nix:298]: not sure if I agree, whitespace will cause this to be mis-interpreted

**Comment by SuperSandro2000** [comment]
I can't test if the change actually works but it does not look to bad.

File: pkgs/games/steam/fhsenv.nix (hunk 1 above)
Comment (jonringer) [reply pkgs/games/steam/fhsenv.nix]: applied

**Comment by compactcode** [comment]
I believe I'm hitting this issue and would like to test this fix out. I'm new to testing out nixpkgs patches, would you be able to point me toward some info on how to check out and test this patch on my local nixos(unstable) machine?

**Comment by TredwellGit** [comment]
@compactcode:
```
git clone https://github.com/NixOS/nixpkgs.git
cd nixpkgs
git fetch origin pull/130984/head:steam-vulkan-fix-icd
git checkout steam-vulkan-fix-icd
git rebase origin/master
nix-shell -I nixpkgs=. -p steam --run steam
```

**Comment by jonringer** [comment]
I should probably check in `/run/opengl-drivers/` first, as that will likely contain nvidia drivers as well, the existing ICD files that get mounted at `/usr/share/vulkan/` are coming from the mesa package, which wouldn't include them

**Comment by jonringer** [comment]
by default, /usr/share/vulkan/icd.d/ will always be populated with mesa, other PR should be preferred

**Comment by jonringer** [comment]
I'm not a bug fan of his, going to go with https://github.com/NixOS/nixpkgs/pull/135024 approach

**Pull request closed without merging**